### PR TITLE
[SPARK-19905][SQL] Bring back Dataset.inputFiles for Hive SerDe tables

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

`Dataset.inputFiles` works by matching `FileRelation`s in the query plan. In Spark 2.1, Hive SerDe tables are represented by `MetastoreRelation`, which inherits from `FileRelation`. However, in Spark 2.2, Hive SerDe tables are now represented by `CatalogRelation`, which doesn't inherit from `FileRelation` anymore, due to the unification of Hive SerDe tables and data source tables. This change breaks `Dataset.inputFiles` for Hive SerDe tables.

This PR tries to fix this issue by explicitly matching `CatalogRelation`s that are Hive SerDe tables in `Dataset.inputFiles`. Note that we can't make `CatalogRelation` inherit from `FileRelation` since not all `CatalogRelation`s are file based (e.g., JDBC data source tables).

## How was this patch tested?

New test case added in `HiveDDLSuite`.

### CODE REVIEW

**Comment by liancheng** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2734,6 +2735,8 @@ class Dataset[T] private[sql](
         fsBasedRelation.inputFiles
       case fr: FileRelation =>
         fr.inputFiles
+      case r: CatalogRelation if DDLUtils.isHiveTable(r.tableMeta) =>
+        r.tableMeta.storage.locationUri.map { _.toString }.toArray
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Didn't use `r.tableMeta.location` here intentionally for safty.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 2]
```diff
@@ -1865,4 +1865,12 @@ class HiveDDLSuite
       }
     }
   }
+
+  test("SPARK-19905: Hive SerDe table input paths") {
+    withTable("spark_19905") {
+      spark.range(10).createOrReplaceTempView("spark_19905_view")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: `withTempView`

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nit: `xx.map(_.toString)`

**Comment by cloud-fan** [comment]
LGTM

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 3]
```diff
@@ -1865,4 +1865,12 @@ class HiveDDLSuite
...
+      sql("CREATE TABLE spark_19905 STORED AS RCFILE AS SELECT * FROM spark_19905_view")
+      assert(spark.table("spark_19905").inputFiles.nonEmpty)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Also try `sql("SELECT input_file_name() FROM spark_19905")`?

Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: `input_file_name` and `Dataset.inputFiles` are different code paths.

**Comment by gatorsmile** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**